server/src/main/java/org/elasticsearch/action/admin/cluster/migration/TransportGetFeatureUpgradeStatusAction.java

```diff
@@ -60,16 +64,50 @@ public TransportGetFeatureUpgradeStatusAction(
     @Override
     protected void masterOperation(Task task, GetFeatureUpgradeStatusRequest request, ClusterState state,
                                    ActionListener<GetFeatureUpgradeStatusResponse> listener) throws Exception {
-        List<GetFeatureUpgradeStatusResponse.IndexVersion> indexVersions = new ArrayList<>();
-        indexVersions.add(new GetFeatureUpgradeStatusResponse.IndexVersion(".security-7", "7.1.1"));
-        List<GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus> features = new ArrayList<>();
-        features.add(new GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus(
-            "security",
-            "7.1.1",
-            "UPGRADE_NEEDED",
+
+        List<GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus> features = systemIndices.getFeatures().entrySet().stream()
+            .map(entry -> getFeatureUpgradeStatus(state, entry))
+            .collect(Collectors.toList());
+
+        boolean isUpgradeNeeded = features.stream()
+                .map(GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus::getMinimumIndexVersion)
+                .min(Version::compareTo)
+                .orElse(Version.CURRENT)
+                .before(Version.V_7_0_0);
+
+        listener.onResponse(new GetFeatureUpgradeStatusResponse(features, isUpgradeNeeded ? "UPGRADE_NEEDED" : "NO_UPGRADE_NEEDED"));
+    }
+
+    private GetFeatureUpgradeStatusResponse.FeatureUpgradeStatus getFeatureUpgradeStatus(
```

Comment:
Could you add a couple unit tests for this method? I like the integration tests in this PR, but none of the test the path that ends up returning `UPGRADE_NEEDED` - doesn't have to be terribly involved, just a quick check is good enough I think.